src/Hamiltonians/abstract.jl

```diff
@@ -129,3 +137,152 @@ julia> rayleigh_quotient(mom, v) # momentum expectation value for state vector `
 ```
 """
 momentum
+
+"""
+    sm, basis = build_sparse_matrix_from_LO(ham::AbstractHamiltonian, add; nnzs = 0)
+
+Create a sparse matrix `sm` of all reachable matrix elements of a linear operator `ham`
+starting from the address `add`. The vector `basis` contains the addresses of basis
+configurations.
+Providing the number `nnzs` of expected calculated matrix elements may improve performance.
+
+See [`BasisSetRep`](@ref).
+"""
+function build_sparse_matrix_from_LO(
+    ham::AbstractHamiltonian, fs=starting_address(ham); nnzs = 0
+)
+    adds = [fs] # list of addresses of length linear dimension of matrix
+    I = Int[]         # row indices, length nnz
+    J = Int[]         # column indices, length nnz
+    V = eltype(ham)[] # values, length nnz
+    if nnzs > 0
+        sizehint!(I, nnzs)
+        sizehint!(J, nnzs)
+        sizehint!(V, nnzs)
+    end
+
+    i = 0 # 1:dim, column of matrix
+    while true # loop over columns of the matrix
+        i += 1 # next column
+        i > length(adds) && break
+        add = adds[i] # new address from list
+        # compute and push diagonal matrix element
+        melem = diagonal_element(ham, add)
+        push!(I, i)
+        push!(J, i)
+        push!(V, melem)
+        for (nadd, melem) in offdiagonals(ham, add) # loop over rows
+            j = findnext(a -> a == nadd, adds, 1) # find index of `nadd` in `adds`
```

Comment:
```suggestion
        for (nadd, melem) in offdiagonals(ham, add) # loop over rows
            iszero(melem) && continue
            j = findnext(a -> a == nadd, adds, 1) # find index of `nadd` in `adds`
```
Adding this here will avoid problems I had with matrices becoming bigger than they have to be. 

I also have some optimizations for this function on my branch (using a `Dict` for indices instead of using `findnext`) which improves performance quite a bit. Should I add it to this PR?

Reply:
Yes, sure. That sounds like a great improvement. Please add it!